src/Bicep.Core/Semantics/DeclaredTypeSymbol.cs

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.Semantics;
+
+public class DeclaredTypeSymbol : DeclaredSymbol
+{
+    public DeclaredTypeSymbol(ISymbolContext context, string name, TypeDeclarationSyntax declaringSyntax, SyntaxBase value)
+        : base(context, name, declaringSyntax, declaringSyntax.Name)
+    {
+        this.Value = value;
+    }
+
+    public TypeDeclarationSyntax DeclaringType => (TypeDeclarationSyntax)this.DeclaringSyntax;
+
+    public SyntaxBase Value { get; }
+
+    public override void Accept(SymbolVisitor visitor) => visitor.VisitDeclaredTypeSymbol(this);
+
+    public override SymbolKind Kind => SymbolKind.Type;
+
+    public override IEnumerable<Symbol> Descendants
```

Comment:
This is hurting my head a little bit, but conceptually, what is the `Type` of a `DeclaredTypeSymbol`?

If `Type` is a self-reference, then I don't think it makes sense to return via the `Descendents` property. I'm also wondering if this is the cause of the recursion you had to add guards for in `SymbolVisitor`.

`Descendents` is a bit loosely-defined, but I'm not sure it makes sense to allow recursion generally in this field - a bit like how it wouldn't make sense for syntax children to be recursive.

Reply:
Well, given a declaration like `type myString = string`, the type of the whole statement is `string`. Maybe this should be wrapped in some kind of indirection, e.g., `type<string>`? You can't use a type statement as an expression, though, so this may be overkill.

Even if there is an extra layer, though, the descendant of a `type<string>` type would be `string`. If the type declared in a type declaration is *not* accessible via the symbol tree, there would need to be an alternate method of surfacing diagnostics included in `ErrorType`s. That will need to visit the whole type symbol tree, which may be recursive.

Reply:
Similar to what you're suggesting, but what about just creating a built-in type (`class TypeType : TypeSymbol`) - that way in a declaration like `type myString = string`, the type of `myString` is just `TypeType`.

As you mention, the user shouldn't ever need to do anything meaningful with this type - it purely exists because our code forces a `Symbol` to have a `Type` (plus it's useful to be able to surface errors by setting the `Type` to `ErrorType`).

I'm probably not thinking something through, but does this block anything? To solve the visiting recursion issue, you can still returned the assigned type for the syntax via the `Descendents` property.

Reply:
I took another look at how we handle other declaration statements, and that having the `DeclaredTypeSymbol` created by `type myString = string` have a type of `string` is perfectly consistent with parameter symbols and type symbols. We use `DeclaredSymbol.Type` for hovers and completions, so using a different property to get the value of a type declaration would require special casing `DeclaredTypeSymbols` wherever we use a symbol to perform an assignability check.

I think maybe everything is just confusingly named? A `DeclaredTypeSymbol` is a `DeclaredSymbol` (like `ParameterSymbol`, `OutputSymbol`, `VariableSymbol`, `ResourceSymbol`, etc) but is not a `TypeSymbol`. It does, however, have a `Kind` of `SymbolKind.Type`, which it probably shouldn't (perhaps a new `SymbolKind.DeclaredType` or `SymbolKind.UserDefinedType` would be more appropriate?).